libsel4vmmplatsupport/src/arch/arm/guest_image.c

```diff
@@ -145,52 +152,57 @@ static int guest_write_address(vm_t *vm, uintptr_t paddr, void *vaddr, size_t si
         int error = seL4_ARM_Page_CleanInvalidate_Data(cap, 0, PAGE_SIZE_4K);
         ZF_LOGF_IFERR(error, "seL4_ARM_Page_CleanInvalidate_Data failed");
     }
-    return 0;
+    return len;
+}
+
+static size_t get_file_size(const char *image_name)
+{
+    FILE *file = fopen(image_name, "r");
+    if (!file) {
+        return 0;
+    }
+    fseek(file, 0, SEEK_END);
+    size_t size = ftell(file);
+    fseek(file, 0, SEEK_SET);
+
+    fclose(file);
+
+    return size;
 }
 
 static int load_image(vm_t *vm, const char *image_name, uintptr_t load_addr,  size_t *resulting_image_size)
 {
-    int fd;
+    int fd[1];
     size_t len;
     int error;
-    fd = open(image_name, 0);
-    if (fd == -1) {
-        ZF_LOGE("Error: Unable to find image \'%s\'", image_name);
+
+    size_t file_size = get_file_size(image_name);
+    if (0 == file_size) {
+        ZF_LOGE("Error: \'%s\' has zero size", image_name);
         return -1;
     }
 
-    /* Try and load the image 1MiB at a time. Reduce the size by half if
-     * there isn't enough memory available.  The total loading time may be
-     * faster if a larger buffer is used as it allows for more batching.
-     */
-    size_t read_size = BIT(20);
-    char *buf = malloc(read_size);
-    while (buf == NULL) {
-        read_size /= 2;
-        if (read_size < 4096) {
-            ZF_LOGE("Not enough memory for copy buffer");
-        }
-        buf = malloc(read_size);
+    fd[0] = open(image_name, 0);
+    if (-1 == fd[0]) {
+        ZF_LOGE("Error: Unable to find image \'%s\'", image_name);
+        return -1;
     }
+
+    vm_ram_mark_allocated(vm, load_addr, ROUND_UP(file_size, PAGE_SIZE_4K));
+
     size_t offset = 0;
-    while (1) {
-        /* Load the image */
-        len = read(fd, buf, read_size);
-        if (!len) {
-            break;
-        }
-        vm_ram_mark_allocated(vm, load_addr + offset, len);
-        error = vm_ram_touch(vm, load_addr + offset, len, guest_write_address, (void *)buf);
-        if (error) {
+    while (offset < file_size) {
+        /* Will return bytes copied to guest memory, or -1 on error */
+        len = vm_ram_touch(vm, load_addr + offset, PAGE_SIZE_4K, guest_write_address, (void *)fd);
```

Comment:
I recall that this function is supposed to support large lengths and it then calls the callback several times, usually one for each page. It should be possible to install a ram device with the largest pages possible and then have this run once for each page?

Reply:
Yes, you're right